hypnotoad/utils/options.py

```diff
@@ -0,0 +1,491 @@
+                        4, doc="option d", value_type=int, allowed=[4, 5, 6]
+                    )
+                    e=WithMeta(
+                        lambda options: options.a + options.b + options.c + options.d,
+                        doc="option e",
+                        value_type=float,
+                        checks=lambda x: x > 0.0
+                    )
+                )
+        """
+        self.__default_values = {key: WithMeta(value) for key, value in kwargs.items()}
+
+        # Add defaults from *args
+        for a in args:
+            for key, value in a.items():
+                if key in self.__default_values:
+                    if value != self.__default_values[key]:
+                        raise ValueError(
+                            f"{key} has been passed more than once with different values"
+                        )
+            self.__default_values.update(
+                {key: WithMeta(value) for key, value in a.items()}
+            )
+
+        # Can be flipped (in create()) to True to allow values to be accessed using
+        # __getitem__ or __getattr__
+        self.__allow_access = False
+
+    @property
+    def defaults(self):
+        """Get the default values defined for this OptionsFactory
+
+        """
+        return deepcopy(self.__default_values)
+
+    def add(self, **kwargs):
+        """Create a more specific version of the factory with extra options. For example,
+        may be useful for a subclass like
+
+            class Parent:
+                options_factory = OptionsFactory(...)
+
+            class Child:
+                options_factory = Parent.options_factory.add(
+                    an_extra_option="used only by Child"
+                )
+
+        Parameters
+        ----------
+        **kwargs : key=[WithMeta, value or expression]
+            The new options to add, these override the ones in the parent factory if key
+            already exists, but keep the doc, allowed and checks if the option is just a
+            new value/expression (not a new WithMeta)
+        """
+        new_default_values = deepcopy(self.__default_values)
+        for key, value in kwargs.items():
+            if (not isinstance(value, WithMeta)) and key in new_default_values:
+                # just update the default value or expression
+                new_default_values[key].value = value
+            else:
+                new_default_values[key] = value
+
+        return OptionsFactory(new_default_values)
+
+    def __getitem__(self, key):
+        if not self.__allow_access:
+            raise AttributeError("Cannot access values from an OptionsFactory")
+
+        # If key is already in __values, then it has a definite value
+        try:
+            return self.__values[key]
+        except KeyError:
+            pass
+
+        # When setting default values, detect circular definitions
+        try:
+            self.__key_chain
+        except AttributeError:
```

Comment:
```suggestion
        if hasattr(self, "__key_chain"):
```
possibly?

The statement by itself looks a bit strange, but is this because the `__` mangles the name?

Reply:
Yeah, I think the `__` mangles the name to `_OptionsFactory__key_chain` so we could do `hasattr(self, "__key_chain")`. I wasn't sure at the time whether the mangling was standard and so could be relied on, but it turned out I had to assume it was in other places as well, so using `hasattr` here is probably better. Will update...